File: templates/matlab/smith_chart.m
function fig = smith_chart()
    addpath(fullfile(fileparts(mfilename('fullpath')),'..','..','_utils','matlab'));
    apply_theme();
    theta = linspace(0, 2*pi, 360);
    fig = figure('Position',[100 100 650 650]);
    plot(cos(theta), sin(theta), 'k', 'LineWidth', 1); hold on;
    rs = [0.2 0.5 1.0 2.0 5.0];
    for k = 1:numel(rs)
        r = rs(k); cx = r/(r+1); R = 1/(r+1);
        plot(cx + R*cos(theta), R*sin(theta), 'Color', [0.55 0.55 0.55], 'LineWidth', 0.7);
        text(cx + R - 0.02, 0.02, sprintf('%g',r), 'FontSize', 7);
    end
    xs = [0.2 0.5 1 2 5 -0.2 -0.5 -1 -2 -5];
    for k = 1:numel(xs)
        x = xs(k); cx = 1; cy = 1/x; R = abs(1/x);
        cxs = cx + R*cos(theta); cys = cy + R*sin(theta);
        mask = cxs.^2 + cys.^2 <= 1 + 1e-9;
        plot(cxs(mask), cys(mask), 'Color', [0.75 0.75 0.75], 'LineWidth', 0.6);
    end
    plot([-1 1], [0 0], 'k', 'LineWidth', 0.8);
    axis equal off; xlim([-1.1 1.1]); ylim([-1.1 1.1]);
    title('Smith chart');
end

File: templates/python/smith_chart.py
"""smith_chart: Smith 阻抗圆图基底."""
import sys, os; sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', '..', '_utils', 'python'))
import numpy as np
import matplotlib.pyplot as plt
from theme import apply_theme

def make_figure(title='Smith chart'):
    apply_theme(fig_size=(6.5, 6.5))
    theta = np.linspace(0, 2*np.pi, 360)
    fig, ax = plt.subplots()
    ax.plot(np.cos(theta), np.sin(theta), 'k', linewidth=1)
    for r in [0.2, 0.5, 1.0, 2.0, 5.0]:
        cx, R = r/(r+1), 1/(r+1)
        ax.plot(cx + R*np.cos(theta), R*np.sin(theta), color='#888', linewidth=0.7)
        ax.text(cx + R - 0.02, 0.02, f'{r}', fontsize=7, color='#444')
    for x in [0.2, 0.5, 1.0, 2.0, 5.0, -0.2, -0.5, -1.0, -2.0, -5.0]:
        cx, cy = 1.0, 1/x; R = abs(1/x)
        pts = np.column_stack([cx + R*np.cos(theta), cy + R*np.sin(theta)])
        inside = (pts[:,0]**2 + pts[:,1]**2) <= 1.0 + 1e-9
        if inside.any():
            ax.plot(pts[inside,0], pts[inside,1], color='#bbb', linewidth=0.6)
    ax.plot([-1, 1], [0, 0], 'k', linewidth=0.8)
    ax.set_xlim(-1.1, 1.1); ax.set_ylim(-1.1, 1.1); ax.set_aspect('equal'); ax.axis('off')
    ax.set_title(title)
    fig.tight_layout()
    return fig

if __name__ == '__main__':
    fig = make_figure()
    fig.savefig(__file__.replace('.py', '.png'), dpi=150)
